tsdate/core.py

```diff
@@ -1180,3 +1390,140 @@ def get_dates(
+    vr_post = np.full(ts.num_nodes, np.nan)  # been an error
+    tables = ts.dump_tables()
+
+    if fixed_node_set is None:
+        fixed_node_set = ts.samples()
+    fixed_nodes = np.array(list(fixed_node_set))
+    mn_post[fixed_nodes] = tables.nodes.time[fixed_nodes]
+    vr_post[fixed_nodes] = 0
+
+    assert np.all(posterior.grid_data[:, 0] > 0), "Invalid posterior"
+
+    metadata_array = tskit.unpack_bytes(
+        ts.tables.nodes.metadata, ts.tables.nodes.metadata_offset
+    )
+    for u in posterior.nonfixed_nodes:
+        # TODO: with method posterior.mean_and_var(node_id) this could be
+        # easily combined with posterior_mean_var
+        pars = posterior[u]
+        mn_post[u] = pars[0] / pars[1]
+        vr_post[u] = pars[0] / pars[1] ** 2
+        metadata_array[u] = json.dumps({"mn": mn_post[u], "vr": vr_post[u]}).encode()
+    md, md_offset = tskit.pack_bytes(metadata_array)
+    tables.nodes.set_columns(
+        flags=tables.nodes.flags,
+        time=tables.nodes.time,
+        population=tables.nodes.population,
+        individual=tables.nodes.individual,
+        metadata=md,
+        metadata_offset=md_offset,
+    )
+    ts = tables.tree_sequence()
+    return ts, mn_post, vr_post
+
+
+def variational_dates(
+    tree_sequence,
+    mutation_rate,
+    population_size=None,
+    recombination_rate=None,
+    priors=None,
+    *,
+    expectation_propagation=1,
+    global_prior=True,
+    eps=1e-6,
+    ignore_oldest_root=False,
+    progress=False,
+    cache_inside=False,
+):
+    """
+    TODO update docstring
+
+    Infer dates for the nodes in a tree sequence, returning an array of inferred dates
+    for nodes, plus other variables such as the posteriors object
+    etc. Parameters are identical to the date() method, which calls this method, then
+    injects the resulting date estimates into the tree sequence
+
+    :return: a tuple of ``(mn_post, posteriors, timepoints, eps, nodes_to_date)``.
+        If the "inside_outside" method is used, ``posteriors`` will contain the
+        posterior probabilities for each node in each time slice, else the returned
+        variable will be ``None``.
+    """
+    # TODO: non-contemporary samples must have priors specified: if so, they'll
+    # work fine with this algorithm.
+    for sample in tree_sequence.samples():
+        if tree_sequence.node(sample).time != 0:
+            raise NotImplementedError("Samples must all be at time 0")
+    fixed_nodes = set(tree_sequence.samples())
+
+    assert expectation_propagation > 0
+
+    # Default to not creating approximate priors unless ts has > 1000 samples
+    approx_priors = False
+    if tree_sequence.num_samples > 1000:
+        approx_priors = True
+
+    if priors is None:
+        if population_size is None:
+            raise ValueError(
+                "Must specify population size if priors are not already "
+                "built using tsdate.build_parameter_grid()"
```

Comment:
should this be `tsdate.build_prior_grid()`?

Reply:
`build_prior_grid` is what the original algo uses to make a set of timepoints from prior quantiles. This is separate from `build_parameter_grid` which dumps the prior parameters instead of a set timepoints. Personally, I think keeping them as separate functions is cleaner -- but I don't have a strong opinion.